jax/scipy/optimize/bfgs_minimize.py

```diff
@@ -0,0 +1,246 @@
+                        H_k=None)
+
+    if max_iterations is None:
+        max_iterations = np.inf
+
+    D = x0.shape[0]
+
+    if analytic_initial_hessian:
+        hess = hessian(func, argnums=0)
+        initial_B = hess(x0)
+        initial_H = np.linalg.pinv(initial_B)
+        state = state._replace(nhev=state.nhev + 1)
+    else:
+        initial_H = np.eye(D)
+
+    value_and_grad = jax.value_and_grad(func)
+
+    f_0, g_0 = value_and_grad(x0)
+    state = state._replace(f_k=f_0, g_k=g_0, H_k=initial_H, nfev=state.nfev + 1, ngev=state.ngev + 1,
+                           converged=np.linalg.norm(g_0) < g_tol)
+
+    def body(state):
+        p_k = -np.dot(state.H_k, state.g_k)
+        line_search_results = line_search(value_and_grad, state.x_k, p_k, f_0=state.f_k, g_0=state.g_k)
+        state = state._replace(nfev=state.nfev + line_search_results.nfev,
+                               ngev=state.ngev + line_search_results.ngev,
+                               failed=line_search_results.failed)
+        s_k = line_search_results.a_k * p_k
+        x_kp1 = state.x_k + s_k
+        f_kp1 = line_search_results.f_k
+        g_kp1 = line_search_results.g_k
+        y_k = g_kp1 - state.g_k
+        rho_k = np.reciprocal(np.dot(y_k, s_k))
+
+        sy_k = s_k[:, None] * y_k[None, :]
+        w = np.eye(D) - rho_k * sy_k
+        H_kp1 = np.dot(np.dot(w, state.H_k), w.T) + rho_k * s_k[:, None] * s_k[None, :]
+
+        converged = np.linalg.norm(g_kp1) < g_tol
+
+        state = state._replace(converged=converged,
+                               k=state.k + 1,
+                               x_k=x_kp1,
+                               f_k=f_kp1,
+                               g_k=g_kp1,
+                               H_k=H_kp1
+                               )
+
+        return state
+
+    state = while_loop(
+        lambda state: np.logical_and(np.logical_and(~ state.converged, ~state.failed), state.k < max_iterations),
+        body,
+        state)
+
+    return state
+
+
+def test_minimize():
+    def rosenbrock(x):
+        return np.sum(100. * np.diff(x) ** 2 + (1. - x[:-1]) ** 2)
+
+    x0 = np.zeros(2)
+
+    @jax.jit
+    def min_op(x0):
+        result = minimize(rosenbrock, x0, analytic_initial_hessian=True)
+        return result
+
+    jax_res1 = min_op(x0)
+
+    print("Final with analytic initialisation:\n", jax_res1)
+
+    @jax.jit
+    def min_op(x0):
+        result = minimize(rosenbrock, x0, analytic_initial_hessian=False)
+        return result
+
+    jax_res2 = min_op(x0)
+
+    print("Final eye initialisation (like scipy):\n", jax_res2)
+
+    from scipy.optimize import minimize as smin
+    import numpy as onp
+
+    def rosenbrock(x):
+        return onp.sum(100. * onp.diff(x) ** 2 + (1. - x[:-1]) ** 2)
+
+    scipy_res = smin(rosenbrock, x0, method='BFGS')
+    print("Scipy:\n", scipy_res)
+
+    assert np.all(np.isclose(scipy_res.x, jax_res1.x_k))
+    assert np.all(np.isclose(scipy_res.x, jax_res2.x_k))
```

Comment:
Are you using the exact same line search procedure as SciPy? If so, would it be possible to compare the exact result after a handful of optimization steps?

I did this for scipy.sparse.linalg.cg and it was very helpful for giving me confidence that I had implemented the algorithm correctly.

Reply:
I'm not using the exact same same line search. I'm using a modified version of backtracking.
Instead of retracting the step size length when the strong Wolfe conditions are not satisfied, we check the sign of u = del_t restricted_func(t). If u > 0 then we follow normal retraction of 0.8, otherwise we expand by 1.2. See fig 3.1 in https://pages.mtu.edu/~struther/Courses/OLD/Sp2013/5630/Jorge_Nocedal_Numerical_optimization_267490.pdf 
and accompanying text. The simple backtracking always results in satisfying W1 but not always strong-W2 because it may be further ahread. With jax it's fairly cheap to check whether we should look ahead or behind for a stationary point. Also, when combined with using the analytic hessian for the initial H_k estimate it often the line search only uses a single iteration.